From 6c08b8af4a48a0cdece12e1edbc17be1f8969901 Mon Sep 17 00:00:00 2001
From: Hongxu Jia <hongxu.jia@windriver.com>
Date: Tue, 29 Oct 2019 17:04:33 +0800
Subject: [PATCH] deploy: Handle a read-only /boot/efi

In order to make `admin unlock --hotfix` work, refer upstream commit
[8894bb3 deploy: Handle a read-only /boot][1], do the same thing on
/boot/efi

[1] https://github.com/ostreedev/ostree/commit/8894bb39498267f4ae06badc7aa54c4eb4bb7f73

Upstream-Status: Inappropriate [OE specific]

Signed-off-by: Hongxu Jia <hongxu.jia@windriver.com>
---
 src/libostree/ostree-sysroot-deploy.c | 26 ++++++++++++++++++++++++++
 1 file changed, 26 insertions(+)

diff --git a/src/libostree/ostree-sysroot-deploy.c b/src/libostree/ostree-sysroot-deploy.c
index a19f81a..63ffa68 100644
--- a/src/libostree/ostree-sysroot-deploy.c
+++ b/src/libostree/ostree-sysroot-deploy.c
@@ -2415,9 +2415,14 @@ ostree_sysroot_write_deployments_with_options (OstreeSysroot     *self,
   else
     {
       gboolean boot_was_ro_mount = FALSE;
+      gboolean bootefi_was_ro_mount = FALSE;
+
       if (self->booted_deployment)
         boot_was_ro_mount = is_ro_mount ("/boot");
 
+      if (self->booted_deployment)
+        bootefi_was_ro_mount = is_ro_mount ("/boot/efi");
+
       g_debug ("boot is ro: %s", boot_was_ro_mount ? "yes" : "no");
 
       if (boot_was_ro_mount)
@@ -2427,6 +2432,13 @@ ostree_sysroot_write_deployments_with_options (OstreeSysroot     *self,
             return glnx_throw_errno_prefix (error, "Remounting /boot read-write");
         }
 
+      g_debug ("boot/efi is ro: %s", bootefi_was_ro_mount ? "yes" : "no");
+      if (bootefi_was_ro_mount)
+        {
+          if (mount ("/boot/efi", "/boot/efi", NULL, MS_REMOUNT | MS_SILENT, NULL) < 0)
+            return glnx_throw_errno_prefix (error, "Remounting /boot/efi read-write");
+        }
+
       if (!_ostree_sysroot_query_bootloader (self, &bootloader, cancellable, error))
         return FALSE;
       bootloader_is_atomic = bootloader != NULL && _ostree_bootloader_is_atomic (bootloader);
@@ -2447,6 +2459,20 @@ ostree_sysroot_write_deployments_with_options (OstreeSysroot     *self,
               g_printerr ("warning: Failed to remount /boot read-only: %s\n", strerror (errno));
             }
         }
+
+      if (bootefi_was_ro_mount)
+        {
+          if (mount ("/boot/efi", "/boot/efi", NULL, MS_REMOUNT | MS_RDONLY | MS_SILENT, NULL) < 0)
+            {
+              /* Only make this a warning because we don't want to
+               * completely bomb out if some other process happened to
+               * jump in and open a file there.  See above TODO
+               * around doing this in a new mount namespace.
+               */
+              g_printerr ("warning: Failed to remount /boot/efi read-only: %s\n", strerror (errno));
+            }
+        }
+
       if (!success)
         return FALSE;
     }
-- 
2.7.4

